ui: Node List Improvements

mrtracy opened the pull request:
The "Node Totals" on the cluster overview page, which links to the Node List,
now provides a detailed breakdown whenever any nodes in the cluster are in the
suspect or dead state.
Resolves #5303
Resolves #13895

A "Node ID" column has been added to the tables on the Node List page; this
allows us to disambiguate between nodes with the same address, an apparently
common occurrence in early clusters.
Resolves #12587

Finally, the cutoff for marking a node in the "Suspect" state has been changed;
rather than an arbitrary 1-minute timeout for expiration, nodes are now
considered suspect immediately if their liveness record is expired. This allows
for much more rapid updates in the UI when a node is take offline, something
that will improve our demoability considerably. This does not fix a specific
issue, but is something I have been considering since adding the suspect
timeout.

![screen shot 2017-03-22 at 4 03 08 pm](https://cloud.githubusercontent.com/assets/6969858/24377147/535844da-130c-11e7-96ec-38a960af39c3.png)

![screen shot 2017-03-27 at 4 25 20 pm](https://cloud.githubusercontent.com/assets/6969858/24377155/56134f08-130c-11e7-8187-60452625ef78.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14395)
<!-- Reviewable:end -->

Commit:
ui: Node List Improvements

The "Node Totals" on the cluster overview page, which links to the Node List,
now provides a detailed breakdown whenever any nodes in the cluster are in the
suspect or dead state.
Resolves #5303
Resolves #13895

A "Node ID" column has been added to the tables on the Node List page; this
allows us to disambiguate between nodes with the same address, an apparently
common occurrence in early clusters.
Resolves #12587

Finally, the cutoff for marking a node in the "Suspect" state has been changed;
rather than an arbitrary 1-minute timeout for expiration, nodes are now
considered suspect immediately if their liveness record is expired. This allows
for much more rapid updates in the UI when a node is take offline, something
that will improve our demoability considerably. This does not fix a specific
issue, but is something I have been considering since adding the suspect
timeout.

tamird commented:
Needs a rebase.

---

Reviewed 2 of 6 files at r1.
Review status: 2 of 6 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-:-KgHCA4JdOmUDrcfAyJf:b-dw4aal)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Rebased with proto changes

tamird commented:
Mostly science dog - would be nice to see screenshots. Didn't review CSS.

---

Reviewed 1 of 6 files at r1, 4 of 4 files at r2.
Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/components/summaryBar.tsx, line 119 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-KgLzFd7_fkLjUScE3QZ:-KgLzFd8r-iT2iUKCv76:b-jr9f4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e269f92cfa46b3fda06ca9dd94cd94a01c2834e/pkg/ui/src/components/summaryBar.tsx#L119)):*
> ```XML
> 
> function SummaryMetricStatHelper(props: MetricsDataComponentProps & SummaryStatProps & { children?: React.ReactNode }) {
>   let datapoints = props.data && props.data.results && props.data.results[0] && props.data.results[0].datapoints;
> ```

consts?

---

*[pkg/ui/src/components/summaryBar.tsx, line 122 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-KgLzIFAJzOSSg-gVSpm:-KgLzIFAJzOSSg-gVSpn:bkoqrc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e269f92cfa46b3fda06ca9dd94cd94a01c2834e/pkg/ui/src/components/summaryBar.tsx#L122)):*
> ```XML
>   let value = datapoints && datapoints[0] && _.last(datapoints).value;
>   let {title, format} = props;
>   return <SummaryStat title={title} format={format} value={_.isNumber(value) ? value : props.value} />;
> ```

what's with this `isNumber` check? is that a null check?

---

*[pkg/ui/src/containers/nodeGraphs.tsx, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-KgLzPzLg5K4ECK1lhW3:-KgLzPzLg5K4ECK1lhW4:bklst3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e269f92cfa46b3fda06ca9dd94cd94a01c2834e/pkg/ui/src/containers/nodeGraphs.tsx#L40)):*
> ```XML
>     }
>     let { nodeCounts } = this.props.nodesSummary.nodeSums;
>     let children: any;
> ```

why any? It's a react element, isn't it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-:-KgLzeuiCN0SGpjAbPnM:b-ww7txx)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
There are screenshots of the two visual changes in the PR description, are they not visible?

tamird commented:
Ah, they are; must've missed 'em in reviewable.

mrtracy commented:
Review status: 3 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/ui/src/components/summaryBar.tsx, line 119 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-KgLzFd7_fkLjUScE3QZ:-KgM0jDJAloKNtn2smFt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e269f92cfa46b3fda06ca9dd94cd94a01c2834e/pkg/ui/src/components/summaryBar.tsx#L119)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

consts?
</blockquote></details>

Done.

---

*[pkg/ui/src/components/summaryBar.tsx, line 122 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-KgLzIFAJzOSSg-gVSpm:-KgM05LsB9xOr79kDjoO:b6dioou) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e269f92cfa46b3fda06ca9dd94cd94a01c2834e/pkg/ui/src/components/summaryBar.tsx#L122)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's with this `isNumber` check? is that a null check?
</blockquote></details>

Yes, this is a null check.  I can't use `if(value)` because zero evaluates to false but is a valid numeric value to display.  `_.isNumber()` is the most succinct check available; it allows zero, but guards against both `null` *and* `undefined` which are unfortunately distinct false states.

---

*[pkg/ui/src/containers/nodeGraphs.tsx, line 40 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-KgLzPzLg5K4ECK1lhW3:-KgM0vsYgC5mbnwA1_5J:b-a9cfgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e269f92cfa46b3fda06ca9dd94cd94a01c2834e/pkg/ui/src/containers/nodeGraphs.tsx#L40)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why any? It's a react element, isn't it?
</blockquote></details>

It's a ReactNode, which is either an Element or Text or some array containing elements or text. Correct, but always very confusing.

Changed to ReactNode, also changed the line above to `const`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/ui/src/components/summaryBar.tsx, line 122 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-KgLzIFAJzOSSg-gVSpm:-KgM4080loWoYmTmlTrS:bt6r9e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e269f92cfa46b3fda06ca9dd94cd94a01c2834e/pkg/ui/src/components/summaryBar.tsx#L122)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Yes, this is a null check.  I can't use `if(value)` because zero evaluates to false but is a valid numeric value to display.  `_.isNumber()` is the most succinct check available; it allows zero, but guards against both `null` *and* `undefined` which are unfortunately distinct false states.
</blockquote></details>

Yuck. Thanks for explaining.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395)*
<!-- Sent from Reviewable.io -->

benesch commented:
It strikes me as somewhat odd that the status indicator is associated with the second column in the table rather than the first. I guess flipping the order of the ID and address columns is equally weird, as is associating the status indicator with the ID, since it's associated with the address in the graph above. ¯\_(ツ)_/¯  

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 6 files at r1, 2 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-:-KgM5KuOxjm1VI6qe39K:b-6g1ind)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
The status indicator used to be its own sortable column, but was changed to the current situation in the most recent redesign.  I think the right thing to do would be to log a  "suggestion" issue in github for this, and maybe give it the "ux-surprise" tag.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14395#-:-KgQOQhrataeXJXQKllB:bbx34ml)*
<!-- Sent from Reviewable.io -->
